extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,202 @@
+                                    RTC_ERROR_ENUM.UNKNOWN_VENDOR, vendor);
+    }
+    // The ad network defined macros override vendor defined/pub specifed.
+    const macros = Object.assign(rtcConfig['vendors'][vendor], customMacros);
+    inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray, rtcStartTime,
+        macros, rtcConfig['timeoutMillis'],
+        vendor);
+  });
+  return Promise.all(promiseArray);
+}
+
+/**
+ * @param {!AMP.BaseElement} a4aElement
+ * @param {!string} url
+ * @param {!Object<string, boolean>} seenUrls
+ * @param {!Array<Promise>} promiseArray
+ * @param {!number} rtcStartTime
+ * @param {!Object} macros
+ * @param {!number} timeoutMillis
+ * @param {string=} opt_vendor
+ * @return
+ */
+function inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray,
+                            rtcStartTime, macros, timeoutMillis,
+                            opt_vendor) {
+  const win = a4aElement.win;
+  const ampDoc = a4aElement.getAmpDoc();
+  if (seenUrls.length == MAX_RTC_CALLOUTS) {
+    return logAndAddErrorResponse(
+        promiseArray, RTC_ERROR_ENUM.MAX_CALLOUTS_EXCEEDED,
+        opt_vendor || url);
+  }
+  const urlReplacements = Services.urlReplacementsForDoc(ampDoc);
+  const whitelist = {};
+  Object.keys(macros || {}).forEach(key => whitelist[key] = true);
+  url = urlReplacements.expandSync(
+      url, macros, undefined, whitelist);
+  if (!isSecureUrl(url)) {
+    return logAndAddErrorResponse(promiseArray, RTC_ERROR_ENUM.INSECURE_URL,
+        opt_vendor || url);
+  }
+  if (!!seenUrls[url]) {
+    return logAndAddErrorResponse(promiseArray, RTC_ERROR_ENUM.DUPLICATE_URL,
+        opt_vendor || url);
+  }
+  seenUrls[url] = true;
+  promiseArray.push(sendRtcCallout_(
+      url, rtcStartTime, win, timeoutMillis, opt_vendor || url));
+}
+
+/**
+ * @param {!string} url
+ * @param {!number} rtcStartTime
+ * @param {!Window} win
+ * @param {!number} timeoutMillis
+ * @param {!string} callout
+ */
+function sendRtcCallout_(url, rtcStartTime, win, timeoutMillis, callout) {
+  /**
+   * Note: Timeout is enforced by timerFor, not the value of
+   *   rtcTime. There are situations where rtcTime could thus
+   *   end up being greater than timeoutMillis.
+   */
+  let rtcTime;
+  return Services.timerFor(win).timeoutPromise(
+      timeoutMillis,
+      Services.xhrFor(win).fetchJson(
+          url, {credentials: 'include'}).then(res => {
+            return res.text().then(text => {
+              rtcTime = Date.now() - rtcStartTime;
+              // An empty text response is allowed, not an error.
+              if (!text) {
+                return {rtcTime, callout};
+              }
+              const rtcResponse = tryParseJson(text);
+              return rtcResponse ? {rtcResponse, rtcTime, callout} :
+              {rtcTime, callout, error: 'Unparsable JSON'};
+            }).catch(unusedError => {
+              return buildErrorResponse(RTC_ERROR_ENUM.BAD_JSON_RESPONSE,
```

Comment:
when would this occur? as part of text()?  I think this falls under NETWORK_FAILURE